Muveto

WenzDaniel opened the pull request:
**What is the problem / what does the code in this PR do**
In this PR we are adding the mveto as a child of the nveto to the processing chain. This PR is accompanied by a technical PR in strax: https://github.com/AxFoundation/strax/pull/355. Travis will fail, but I pytested it on my local machine. 

In addition, I made some smaller changes for the nveto:
* I removed the n_nveto_pmts option since it is not really needed (not like in case of the TPC).  
* The to_pe constants can now be accessed via the CMT.

jorana commented:
@WenzDaniel , let me know when ready. I would appreciate if you un-draft when this is the case.

WenzDaniel commented:
Beside the already mentioned changes. I updated all child plugins according to the changes in https://github.com/AxFoundation/strax/pull/355. Further, I added a test for all child-plugins which tests if options and lineages are updated correctly.

jorana commented on straxen/plugins/veto_hitlets.py:
```diff
@@ -69,23 +70,29 @@ class nVETOHitlets(strax.Plugin):
 
     provides = 'hitlets_nv'
     data_kind = 'hitlets_nv'
+    ends_with = '_nv'
```
Still needed?

WenzDaniel replied:
Oh no good catch.

jorana commented on straxen/plugins/nveto_recorder.py:
```diff
@@ -20,9 +20,7 @@
-                      "channel number."),
-    strax.Option('n_nveto_pmts', type=int, track=False,
-                 help='Number of nVETO PMTs')
+                      "channel number.")
```
extremely minor, if you have a list like this. Ending with `,` will keep the git history cleaner as now it will change this line whereas you did not change anything actually.

WenzDaniel replied:
Will change it

jorana commented on straxen/plugins/veto_hitlets.py:
```diff
@@ -69,23 +70,29 @@ class nVETOHitlets(strax.Plugin):
+        to_pe = straxen.get_to_pe(self.run_id,
+                                  self.config['gain_model_nv'],
+                                  n_tpc_pmts=straxen.n_nveto_pmts)
+        self.channel = self.config['channel_map']['nveto']
```
`self.channel_range`?

jorana commented on straxen/plugins/veto_hitlets.py:
```diff
@@ -159,3 +166,67 @@ def drop_data_field(old_hitlets, new_hitlets):
+                 child_option=True, parent_option_name='gain_model_nv',
+             help='PMT gain model. Specify as (model_type, model_config)'),
+)
+class muVETOHitlets(nVETOHitlets):
```
man, you really nailed it with the `child_plugins`, look at how short this class is!

WenzDaniel replied:
Thanks :)

jorana commented on straxen/plugins/veto_pulse_processing.py:
```diff
@@ -33,6 +40,7 @@ class nVETOPulseProcessing(strax.Plugin):
     depends_on = 'raw_records_coin_nv'
     provides = 'records_nv'
     data_kind = 'records_nv'
+    ends_with = '_nv'
```
Uhm, bit confused as to if this should be like this or True?

jorana commented on straxen/plugins/veto_pulse_processing.py:
```diff
@@ -127,3 +135,42 @@ def clean_up_empty_records(records, record_links, only_last=True):
+    depends_on = 'raw_records_mv'
+    provides = 'records_mv'
+    data_kind = 'records_mv'
+    ends_with = '_mv'
```
Perhaps I don't fully see why this is different from the `_he`?

jorana commented on straxen/plugins/veto_pulse_processing.py:
```diff
@@ -127,3 +135,42 @@ def clean_up_empty_records(records, record_links, only_last=True):
+        return dtype
+
+    def compute(self, raw_records_mv):
+        return super().compute(raw_records_mv)
```
please fix

jorana commented on tests/test_plugins.py:
```diff
@@ -134,6 +137,44 @@ def _update_context(st, max_workers, fallback_gains=None):
             'timeout': 60,  # we don't want to build travis for ever
         })
 
+def _test_child_options(st):
```
Although this looks like a `strax`-like test I think it will do no harm and do catch potential problems :)

jorana reviewed (commented):
Hi Daniel! I think this PR really demonstrates the power of the child plugins. I think we need to be smart here and there so as to not create a thousand datatypes that no one will look at. Can we summarize the plugins that will now be used for mu and nu Veto and check what of these data should be stored long term. I think we should also include an AC @feigaodm here. Saving more will have pros and cons. It's hard to know what makes sense at this stage.

Once discussed, I/you can also add a separate heading in the documentation for the muon veto.

WenzDaniel commented on straxen/plugins/nveto_recorder.py:
```diff
@@ -54,7 +52,7 @@ def infer_dtype(self):
         nveto_records_dtype = strax.raw_record_dtype(self.record_length)
         nveto_diagnostic_lone_records_dtype = strax.record_dtype(self.record_length)
-        nveto_lone_records_statistics_dtype = lone_record_statistics_dtype(self.config['n_nveto_pmts'])
+        nveto_lone_records_statistics_dtype = lone_record_statistics_dtype(straxen.n_nveto_pmts)
```
I have no strong opinion on that. I was also very indecisive. I removed it to reduce the number of options. Bu I agree in general it is better to not hardcode something.

WenzDaniel commented:
Addressed all comments

jorana commented:
Can you also address the written comment?

WenzDaniel commented:
> Can we summarize the plugins that will now be used for mu and nu Veto and check what of these data should be stored long term.

I think during commissioning we will store all data kinds to see if additional adjustments have to be made. In the long run we will keep probably everything except for  `raw_records_coin_nv`. So there is not much to be reduced.

jorana commented:
> I think during commissioning we will store all data kinds to see if additional adjustments have to be made. In the long run we will keep probably everything except for raw_records_nv. So there is not much to be reduced.

I don't see the point in this but perhaps @feigaodm can comment. Why keep all kinds of data that people will never look at.

jorana commented:
by the way, we shouldn't forget upping the requirements for strax to the next releases version directly after merging https://github.com/AxFoundation/strax/pull/355
